### PR TITLE
Handle incorrect arg values in controls

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: #14926

## What I did

Added stories highlighting the problem.

### CODE REVIEW

File: lib/components/src/controls/Boolean.stories.tsx [hunk 1]
```diff
@@ -11,11 +11,18 @@ const Template = (initialValue?: boolean) => {
   return (
     <>
       <BooleanControl name="boolean" value={value} onChange={(newVal) => setValue(newVal)} />
-      <p>value: {typeof value === 'boolean' ? value.toString() : value}</p>
+      <pre>{JSON.stringify(value) || 'undefined'}</pre>
```
Comment (ghengeveld) [inline lib/components/src/controls/Boolean.stories.tsx:14]: `String(value)` would've done the trick. Unless you want additional quotes around strings.

File: lib/components/src/controls/Boolean.tsx [hunk 2]
```diff
@@ -80,16 +81,23 @@ const format = (value: BooleanValue): string | null => (value ? String(value) :
...
-      checked={value || false}
-      {...{ name, onBlur, onFocus }}
-    />
-    <span>False</span>
-    <span>True</span>
-  </Label>
-);
+export const BooleanControl: FC<BooleanProps> = ({ name, value, onChange, onBlur, onFocus }) => {
+  const onSetFalse = useCallback(() => onChange(false), [onChange]);
+  if (value === undefined) {
```
Comment (ghengeveld) [inline lib/components/src/controls/Boolean.tsx:86]: What about `null`?

File: lib/components/src/controls/Number.stories.tsx [hunk 3]
```diff
@@ -6,22 +6,22 @@ export default {
...
-  return (
-    <>
-      <NumberControl name="number" value={value} onChange={(newVal) => setValue(newVal)} />
-      <p>{value}</p>
-    </>
-  );
-};
+export const Basic = () => Template(10);
+
+export const Zero = () => Template(0);
```
Comment (ghengeveld) [inline lib/components/src/controls/Number.stories.tsx:21]: What about decimals?

File: lib/components/src/controls/Range.tsx [hunk 4]
```diff
@@ -27,6 +27,13 @@ const RangeInput = styled.input(({ theme }) => ({
     cursor: 'pointer',
   },
 
+  // FIXME
```
Comment (ghengeveld) [inline lib/components/src/controls/Range.tsx:30]: Would be nice to explain what needs to be fixed here.

File: lib/components/src/controls/Boolean.stories.tsx (hunk 1 above)
Comment (tmeasday) [reply lib/components/src/controls/Boolean.stories.tsx:14]: This is just common to all the stories for the different controls

**Review by ghengeveld — APPROVED** [review]

File: lib/components/src/controls/Boolean.tsx (hunk 2 above)
Comment (tmeasday) [reply lib/components/src/controls/Boolean.tsx:86]: Yes, sorry this PR was supposed to be a draft, and was based on other PRs that handled `undefined.

It doesn't actually handle incorrect arg values yet -- e.g. `null` or the wrong type.

File: lib/components/src/controls/Number.stories.tsx (hunk 3 above)
Comment (tmeasday) [reply lib/components/src/controls/Number.stories.tsx:21]: They work, but sure, we should add a story.

File: lib/components/src/controls/Range.tsx (hunk 4 above)
Comment (tmeasday) [reply lib/components/src/controls/Range.tsx:30]: This isn't really part of this PR. There is a separate open ticket about this exact thing.

**Review by tmeasday — COMMENTED** [review]
Sorry this was a bit confusing @ghengeveld

**Comment by tmeasday** [comment]
Not sure why this PR still includes a bunch of changes that were already merged to `next`. Probably it needs to be rebased.

**Review by ndelangen — APPROVED** [review]

**Comment by tmeasday** [comment]
I guess we can't really merge this until we get the controls to handle the errors at least a little better.

**Comment by domyen** [comment]
@tmeasday Is there anything I need to do here?

**Comment by tmeasday** [comment]
@domyen, yeah, let's figure out a design for what controls look like when they have invalid content.

A simple example is something like

```js
export const MyStory = Template.bind({})
MyStory.argTypes = { foo: { type: 'string' } }
MyStory.args = { foo: 7 }
```

We've said `foo` is a string, but supplied the value `7`. What should we show in the controls UI in this case? (Chances are the actually story will error out).

Another example is when an arg is required but the story does not supply a value.

**Comment by tmeasday** [comment]
@ndelangen -- have discussed this with @shilman and we were more of the opinion that we just let the user do what they want here -- keep in mind that `argTypes` can be inferred from TS/propTypes and TBH I don't entirely trust docgen to always get them right.

So we were thinking the issue is more just how do we deal with trying to render a value with a control that it is not equipped to handle.

**Comment by ndelangen** [comment]
> could choose to render an error rather then the story
OK so that's out of the question then.

How about rendering the warning where the control would be?

**Comment by tmeasday** [comment]
> How about rendering the warning where the control would be?

Seems like an OK solution. Perhaps we want to let the override the "broken" value with the control though?

**Comment by ndelangen** [comment]
Like a button to set it to the default value?

**Comment by tmeasday** [comment]
Sure, I guess the PR isn't helping much unmerged. But FTR I don't think we currently handle all possibilities of weird incorrect things you could pass into controls.

**Pull request closed without merging**